### DISCUSSION TITLE
Overfitting Pemodelan

### DISCUSSION THREAD

**Discussion by joejnt** [teamalgoritma/community#511] [Q&A] [answered]
Dear Tim Algoritma,

Mohon saran dan pendapatnya, apabila kita membuat pemodelan klasifikasi dengan jumlah obv. 120 (split data 20:80), dari hasil pemodelan akurasinya mencapai 1.00, apakah kemungkinan terjadi overfitting? Apabila benar model overfit, treatmentnya seperti apa ya karena sudah tidak bisa menambah jumlah observasi lagi karena itu merupakan data populasi.

Best regards,
JT

**Comment by ramdhinafinita** [comment, marked as answer]
Dear @joejnt 

Terimakasih atas pertanyaannya. 

Ada kemungkinan model memiliki akurasi 1,00 tetapi hal itu masih dikatakan terjadi overfitting, terutama jika ukuran sampelnya relatif kecil dan model yang dibuat kompleks. Salah satu cara untuk menguji overfitting adalah dengan mengevaluasi performa model pada set pengujian data test yang terpisah. Jika model bekerja secara signifikan lebih buruk pada set pengujian daripada pada set pelatihan, kemungkinan overfitting. Karena seharusnya, pembuatan model memiliki 3 dataset yakni train, validation, dan test. Dimana, tahap splitting yang bapak lakukan digunakan untuk membagi data train (80%)  dan validation (20%). 

Untuk mengatasi overfitting, Bapak dapat mencoba menggunakan teknik regularisasi, seperti regularisasi L1 atau L2, yang menambahkan istilah penalti pada model's loss function  untuk mencegah penyesuaian noise pada data. Pendekatan lain adalah dengan menggunakan model yang lebih sederhana, seperti regresi linier atau decission tree, yang memiliki lebih sedikit parameter yang sesuai. Terakhir, Bapak dapat mencoba mengumpulkan lebih banyak data untuk menambah ukuran set pelatihan, jika memungkinkan.

Referensi handling overfitting:
https://towardsdatascience.com/8-simple-techniques-to-prevent-overfitting-4d443da2ef7d
https://medium.datadriveninvestor.com/how-to-handle-overfitting-and-underfitting-470a1f7389fe

Semoga bermanfaat :)

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Terima kasih untuk sarannya, berikut hasil yang saya peroleh pada saat membuat pemodelan klasifikasi menggunakan DT dan RF.

Model DT
<img width="380" alt="image" src="https://user-images.githubusercontent.com/103427989/214263555-3d723a70-4ac9-4eb9-a19b-69938d820883.png">


Model RF
<img width="370" alt="image" src="https://user-images.githubusercontent.com/103427989/214263702-ccfeaa03-0e7c-45be-9dbf-b09372a9c543.png">


Mohon saran dan pendapatnya.

Best regards,
JT

**Reply by ramdhinafinita** [reply]
Dear @joejnt

Benar pak, jika dari hasil model yang sudah bapak berikan itu menandakan terjadinya overfitting. Namun bapak bisa coba melakukan tuning pada model, contohnya melakukan model tuning pada random forest dengan library ranger. 
Menggunakan ranger dari package Random Forest. Tahapan tunning pada model random forest dilakuan untuk mencari parameter terbaik. Langkah-langkah yang dilakukan pada tahap ini yakni sebagai berikut:

1. Membuat kombinasi dari parameter (mtry, min.node.size, sample_size),
2. Menggunakan fungsi ranger untuk mencari parameter terbaik berdasarkan percobaan pembuatan model dengan data train,
3. Mencari parameter yang menghasilkan nilai out-of-bag (OOB) error yang kecil,
4. Membuat model dengan parameter tersebut.

reference:
https://machinelearningmastery.com/tune-machine-learning-algorithms-in-r/
https://rpubs.com/phamdinhkhanh/389752
https://towardsdatascience.com/decision-tree-hyperparameter-tuning-in-r-using-mlr-3248bfd2d88c
https://www.guru99.com/r-decision-trees.html

Berikut yang dapat saya sampaikan, semoga bermanfaat.

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Terima kasih untuk inside nya.

Best regards,
JT

**Reply by muhrisman** [reply]
Izin menambahkan, problem tersebut bisa juga terjadi karena adanya data leakage. Data leakage adalah kondisi dari variabel target yang bisa dijelaskan langsung oleh variabel prediktor. Bisa dicoba untuk menelusuri variabel - variabel dari prediktor yang kemungkinan mengakibatkan kondisi ini.

**Reply by joejnt** [reply]
Dear @muhrisman 

Well noted, terima kasih untuk informasinya. Saya mendapatkan informasi bahwa untuk mencegah data leakage dalam machine learning salah satunya adalah dengan menggunakan teknik kross-validasi: Ini termasuk teknik seperti k-fold cross-validation yang membantu untuk mengevaluasi model dengan **data yang tidak digunakan untuk pelatihan model**. Namun saya masih tidak mengerti maksud dari **data yang tidak digunakan untuk pelatihan model**. Apakah ada referensi untuk metode k-fold cross-validation?

Best regards,
JT

**Reply by muhrisman** [reply]
Data yang tidak digunakan untuk pelatihan model itu maksudnya data test pak. Jadi data yang dipakai di training dievaluasi menggunakan data testing.

Untuk referensi k-fold cross validation ada di materi classification 2 Algoritma.

Berikut referensinya https://www.analyticsvidhya.com/blog/2022/02/k-fold-cross-validation-technique-and-its-essentials/

Untuk k fold validation mudahnya gini, kita punya data utuh, kita bagi jadi data training dan testing. Kita latih modelnya di data training dan evaluasi di data testing.

Kemudian kita lakukan lagi proses pembagian data tersebut namun data-data yang belum sempat mencoba posisi sebagai data testing akan disampling ulang sehingga dimasukan sebagai data training.
Hal tersebut dilakukan sampai semua observasi data berperan sebagai data testing.

Kenapa hal ini dilakukan? Saat kita membagi data testing dan data training kemudian melakukan evaluasi kita dapatkan hasilnya. Lalu bagaimana kalau kita lakukan sampling ulang dengan data testing yang berbeda? Apakah hasilnya berbeda signifikan?
Nah k fold cross validation akan berperan dalam hal ini sehingga semua observasi data memiliki peran sebagai data testing.

Sekian, semoga membantu :) 

Best Regards,
M. Risman
Former Instructor Algoritma 😁

**Reply by joejnt** [reply]
Dear @muhrisman 

Well noted, terima kasih untuk insidenya 👍 

Best regards,
JT
